zkevm-circuits/src/keccak_circuit/keccak_packed.rs

```diff
@@ -0,0 +1,1770 @@
+    bits: Vec<usize>,
+}
+
+/// Part
+#[derive(Clone, Debug)]
+pub(crate) struct WordParts {
+    parts: Vec<PartInfo>,
+    rot_idx: usize,
+}
+
+fn get_word_parts(part_size: usize, rot: usize, normalize: bool) -> WordParts {
+    let mut bits = (0usize..64).collect::<Vec<_>>();
+    bits.rotate_right(rot);
+
+    let mut parts = Vec::new();
+    let mut rot_idx = 0;
+
+    let mut idx = 0;
+    let target_sizes = if normalize {
+        target_part_sizes(part_size)
+    } else {
+        let num_parts_a = rot / part_size;
+        let partial_part_a = rot % part_size;
+
+        let num_parts_b = (64 - rot) / part_size;
+        let partial_part_b = (64 - rot) % part_size;
+
+        let mut part_sizes = vec![part_size; num_parts_a];
+        if partial_part_a > 0 {
+            part_sizes.push(partial_part_a);
+        }
+
+        part_sizes.extend(vec![part_size; num_parts_b]);
+        if partial_part_b > 0 {
+            part_sizes.push(partial_part_b);
+        }
+
+        part_sizes
+    };
+
+    for part_size in target_sizes {
+        let mut num_consumed = 0;
+        while num_consumed < part_size {
+            let mut part_bits: Vec<usize> = Vec::new();
+            while num_consumed < part_size {
+                if !part_bits.is_empty() && bits[idx] == 0 {
+                    break;
+                }
+                if bits[idx] == 0 {
+                    rot_idx = parts.len();
+                }
+                part_bits.push(bits[idx]);
+                idx += 1;
+                num_consumed += 1;
+            }
+            parts.push(PartInfo { bits: part_bits });
+        }
+    }
+    assert_eq!(get_rotate_count(rot, part_size), rot_idx);
+
+    parts.rotate_left(rot_idx);
+    assert_eq!(parts[0].bits[0], 0);
+
+    WordParts { parts, rot_idx }
+}
+
+fn combine_sub_parts<F: Field>(
+    name: &'static str,
+    meta: &mut ConstraintSystem<F>,
+    lookup_counter: &mut usize,
+    input: Vec<Part<F>>,
+    part_size: usize,
+    range_check_table: TableColumn,
+    range_check: bool,
+) -> Vec<Part<F>> {
+    let target_sizes = target_part_sizes(part_size);
+    let mut counter = 0;
+    let mut parts = Vec::new();
+    let mut input_iter = input.iter();
+    while let Some(input_part) = input_iter.next() {
+        if input_part.num_bits == target_sizes[counter] {
+            parts.push(input_part.clone());
+            counter += 1;
+        } else if let Some(extra_part) = input_iter.next() {
+            assert_eq!(
+                input_part.num_bits + extra_part.num_bits,
+                target_sizes[counter]
+            );
+
+            let factor = F::from(8u64).pow(&[input_part.num_bits as u64, 0, 0, 0]);
+            let expr = input_part.expr.clone() + extra_part.expr.clone() * factor;
+
+            // Could do a couple of these together when the parts are small to save some
+            // lookups println!("{} + {}", input_part.num_bits,
+            // extra_part.num_bits);
```

Comment:
Yes exactly what you say. If the tables are able to do 8 bits/lookup, but we have e.g. two small parts of 4 bits we can use a single lookup instead. But yeah didn't do it because that gets a bit crazy.